internal/types/subtypes/interceptor_test.go

```diff
@@ -0,0 +1,575 @@
+							return attrs
+						}(),
+					},
+				},
+			},
+		},
+		{
+			"TestUpdateResourceResponse",
+			&attribute.TestUpdateResourceResponse{
+				Item: &attribute.TestResource{
+					Id:   "trsr_one",
+					Type: "sub_resource",
+					Attrs: &attribute.TestResource_SubResourceAttributes{
+						SubResourceAttributes: &attribute.TestSubResourceAttributes{
+							Name: "test",
+						},
+					},
+				},
+			},
+			&attribute.TestUpdateResourceResponse{
+				Item: &attribute.TestResource{
+					Id:   "trsr_one",
+					Type: "sub_resource",
+					Attrs: &attribute.TestResource_Attributes{
+						Attributes: func() *structpb.Struct {
+							attrs, _ := structpb.NewStruct(map[string]interface{}{
+								"name": "test",
+							})
+							return attrs
+						}(),
+					},
+				},
+			},
+		},
+		{
+			"TestResponseNoItem",
+			&attribute.TestResponseNoItem{
+				Name: "test",
+			},
+			&attribute.TestResponseNoItem{
+				Name: "test",
+			},
+		},
+		{
+			"TestResponseItemNotMessage",
+			&attribute.TestResponseItemNotMessage{
+				Item: "test",
+			},
+			&attribute.TestResponseItemNotMessage{
+				Item: "test",
+			},
+		},
+		{
+			"TestResponseItemNoType",
+			&attribute.TestResponseItemNoType{
+				Item: &attribute.TestItemNoType{
+					Id: "test",
+				},
+			},
+			&attribute.TestResponseItemNoType{
+				Item: &attribute.TestItemNoType{
+					Id: "test",
+				},
+			},
+		},
+		{
+			"TestCreateNoOneOfResponse",
+			&attribute.TestCreateNoOneOfResponse{
+				Item: &attribute.TestNoOneOf{
+					Id:   "trsr_one",
+					Type: "sub_resource",
+					Attributes: func() *structpb.Struct {
+						attrs, _ := structpb.NewStruct(map[string]interface{}{
+							"name": "test",
+						})
+						return attrs
+					}(),
+				},
+			},
+			&attribute.TestCreateNoOneOfResponse{
+				Item: &attribute.TestNoOneOf{
+					Id:   "trsr_one",
+					Type: "sub_resource",
+					Attributes: func() *structpb.Struct {
+						attrs, _ := structpb.NewStruct(map[string]interface{}{
+							"name": "test",
+						})
+						return attrs
+					}(),
+				},
+			},
+		},
+		{
+			"TestUpdateNoOneOfResponse",
+			&attribute.TestUpdateNoOneOfResponse{
+				Item: &attribute.TestNoOneOf{
+					Id:   "trsr_one",
+					Type: "sub_resource",
+					Attributes: func() *structpb.Struct {
+						attrs, _ := structpb.NewStruct(map[string]interface{}{
+							"name": "test",
+						})
+						return attrs
+					}(),
+				},
+			},
+			&attribute.TestUpdateNoOneOfResponse{
+				Item: &attribute.TestNoOneOf{
+					Id:   "trsr_one",
+					Type: "sub_resource",
+					Attributes: func() *structpb.Struct {
+						attrs, _ := structpb.NewStruct(map[string]interface{}{
+							"name": "test",
+						})
+						return attrs
+					}(),
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := transformResponseAttributes(tc.resp)
+			require.NoError(t, err)
+			assert.Empty(t, cmp.Diff(got, tc.expected, protocmp.Transform()))
+		})
+	}
+}
```

Comment:
Should we have an error test as well for these functions? Personally I'd probably argue that we don't necessarily need both the internal and external tests so I'll leave it up to you.

Reply:
I think in this case the external test is more important, since the particular error message and type is important as it will be returned. And that error message is determined in the exported interceptor function. The particular errors from the un-exported functions imo should be free to change.

Reply:
Agreed the external tests are more important in general.